component/mimir/rules/diff.go

```diff
@@ -0,0 +1,111 @@
+package rules
+
+import (
+	"bytes"
+
+	mimirClient "github.com/grafana/agent/pkg/mimir/client"
+
+	"gopkg.in/yaml.v3"
+)
+
+type ruleGroupDiffKind string
+
+const (
+	ruleGroupDiffKindAdd    ruleGroupDiffKind = "add"
+	ruleGroupDiffKindRemove ruleGroupDiffKind = "remove"
+	ruleGroupDiffKindUpdate ruleGroupDiffKind = "update"
+)
+
+type ruleGroupDiff struct {
+	Kind    ruleGroupDiffKind
+	Actual  mimirClient.RuleGroup
+	Desired mimirClient.RuleGroup
+}
+
+func diffRuleState(desired map[string][]mimirClient.RuleGroup, actual map[string][]mimirClient.RuleGroup) map[string][]ruleGroupDiff {
```

Comment:
I tried doing that, but it ends up forcing me to allocate copies of the original maps just to satisfy the types. I wonder if it could be better to just define a type for the whole map, which is easier to just type convert.

Reply:
Maybe just a doc comment on the function then?

Reply:
Oh, a type for the whole map could work too

Reply:
Looks good, thanks!